go/sync2/batcher.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2017 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sync2
+
+import (
+	"time"
+)
+
+// Batcher delays concurrent operations for a configurable interval in order to
+// batch them up or otherwise clock their operation to run concurrently.
+//
+// It is implemented as a channel of int32s. Each waiter blocks on the channel
+// from which it gets a sequentially increasing batch ID when the timer elapses.
+//
+// Hence a waiter is delayed for at most the batch interval.
+type Batcher struct {
+	interval time.Duration
+	queue    chan int32
+	waiters  AtomicInt32
+	nextID   AtomicInt32
+}
+
+// NewBatcher returns a new Batcher
+func NewBatcher(interval time.Duration) *Batcher {
+	return &Batcher{
+		interval: interval,
+		queue:    make(chan int32),
+		waiters:  NewAtomicInt32(0),
+		nextID:   NewAtomicInt32(0),
+	}
+}
+
+// Wait adds a new waiter to the queue and blocks until the next batch
+func (b *Batcher) Wait() int32 {
+	numWaiters := b.waiters.Add(1)
+	if numWaiters == 1 {
+		b.newBatch()
+	}
+	return <-b.queue
+}
+
+// newBatch starts a new batch
+func (b *Batcher) newBatch() {
+	go func() {
+		time.Sleep(b.interval)
```

Comment:
For containers, we've seen cases where a process can lose CPU for long periods of time, in which case this may fail. I'm wondering if `runtime.Gosched` may do a better job, and possibly be faster.

Of course, we'll need to run these tests on a heavily loaded machine. For now, I think this is ok. We may have to revisit this if we see flakyness.

Reply:
Yeah, I also think that this will result into flaky tests. (That's the case when you launch multiple Go routines in your test code and not all of them will finish their time.Sleep(n) before the test finishes its time.Sleep(2*n) and moves on to the next test case. Weird delays of single routines, but not all of them, and therefore a re-ordering can happen on slow machines.)

Our usual trick is to mock time.Now() when it comes to comparing timestamps.

In general, we try to avoid time.Sleep() in test code because it makes the tests slower and flaky.

In this case, you could make time.Sleep() configurable and your test could provide a more sophisticated version for which you control when it unblocks. (Personally, I like to provide a second private constructor which the test can use to inject their implementation. Example for time.Now(): https://github.com/youtube/vitess/blob/master/go/vt/vtgate/buffer/buffer.go#L107)

In your case, your fake Sleep() could block on reading a channel that was created by the test method. When you close the channel in the test method, time.Sleep() unblocks and triggers the rest of your code.

Does that make sense?

For the channel, please use an empty struct since the values are irrelevant and in fact you don't need to put anything into it. Instead, you'll only use the closure of the channel as signal. I looked around and think that this test is a similar example for using such a channel as notification: https://github.com/youtube/vitess/blob/master/go/stats/counters_test.go#L177 We also have actual code overall the place where we use this pattern.

Reply:
I had a long conversation with @sougou on Slack about ways to get rid of the Sleep. Unfortunately there doesn't seem to be any way to do this given how vtgate works, since there's no way to reliably determine when a query execution is blocked waiting for response(s) from the tablet(s) or when it might issue additional queries before waiting.

To that end, I'll try to make the timeouts a bit more forgiving when running in Travis, and if that doesn't work we may have to mark the tests as flaky.